Source/WebCore/dom/ShadowRoot.cpp

```diff
@@ -225,10 +225,20 @@ void ShadowRoot::setResetStyleInheritance(bool value)
     m_resetStyleInheritance = value;
 }
 
-Ref<Node> ShadowRoot::cloneNodeInternal(Document&, CloningOperation)
+Ref<Node> ShadowRoot::cloneNodeInternal(Document& targetDocument, CloningOperation type)
 {
-    RELEASE_ASSERT_NOT_REACHED();
-    return *static_cast<Node*>(nullptr); // ShadowRoots should never be cloned.
+    RELEASE_ASSERT(m_mode != ShadowRootMode::UserAgent);
+    switch (type) {
+    case CloningOperation::SelfWithTemplateContent:
+        return create(targetDocument, m_mode, m_slotAssignmentMode, m_delegatesFocus ? DelegatesFocus::Yes : DelegatesFocus::No,
+            m_cloneable ? Cloneable::Yes : Cloneable::No, m_availableToElementInternals ? AvailableToElementInternals::Yes : AvailableToElementInternals::No);
+    case CloningOperation::OnlySelf:
+    case CloningOperation::Everything:
+        break;
+    }
+
+    RELEASE_ASSERT_NOT_REACHED(); // ShadowRoot is never cloned directly on its own.
+    return static_cast<Node&>(*this);
```

Comment:
I am a little surprised that the static_cast is needed.

Reply:
That's definitely not necessary. Removed.